src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DoNotRaiseExceptionsInUnexpectedLocations.cs

```diff
@@ -1,59 +1,351 @@
+                            {
+                                operationContext.ReportDiagnostic(
+                                    operation.Syntax.CreateDiagnostic(methodCategory.Rule, methodSymbol.Name, type.Name));
+                            }
+                        }
+                    }, OperationKind.ThrowStatement);
+                });
+            });
+        }
+
+        /// <summary>
+        /// This object describes a class of methods where exception throwing statements should be analyzed.
+        /// </summary>
+        private class MethodCategory
+        {
+            /// <summary>
+            /// Function used to determine whether a given method symbol should be analyzed.
+            /// </summary>
+            private readonly Func<IMethodSymbol, Compilation, bool> matchFunction;
             
+            /// <summary>
+            /// Determines if we should analyze non-public methods of a given type.
+            /// </summary>
+            private readonly bool analyzeOnlyPublicMethods;
+
+            /// <summary>
+            /// The rule that should be fired if there is an exception in this kind of method.
+            /// </summary>
+            public DiagnosticDescriptor Rule { get; }
+
+            /// <summary>
+            /// List of exception types which are allowed to be thrown inside this category of method.
+            /// This list will be empty if no exceptions are allowed.
+            /// </summary>
+            public ImmutableArray<ITypeSymbol> AllowedExceptions { get; }
+
+            public MethodCategory(Func<IMethodSymbol, Compilation, bool> matchFunction, bool analyzeOnlyPublicMethods, DiagnosticDescriptor rule, params ITypeSymbol[] allowedExceptionTypes)
+            {
+                this.matchFunction = matchFunction;
+                this.analyzeOnlyPublicMethods = analyzeOnlyPublicMethods;
+                this.Rule = rule;
+                AllowedExceptions = ImmutableArray.Create(allowedExceptionTypes);
+            }
+
+            /// <summary>
+            /// Checks if the given method belong this category
+            /// </summary>
+            public bool IsMatch(IMethodSymbol method, Compilation compilation)
+            {
+                // If we are supposed to analyze only public methods get the resultant visibility
+                // i.e public method inside an internal class is not considered public.
+                if (analyzeOnlyPublicMethods &&
+                    method.GetResultantVisibility() != SymbolVisibility.Public)
+                {
+                    return false;
+                }
+
+                return matchFunction(method, compilation);
+            }
+        }
+        
+        private static List<MethodCategory> GetMethodCategories(Compilation compilation)
+        {
+            var methodCategories = new List<MethodCategory>(12);
```

Comment:
Rather than having to count the categories, consider collection initializer:

```C#
var methodCategories = new List<MethodCategory>
{
    new MethodCategory(IsPropertyGetter, true...),
    new MethodCategory(IsIndexGetter, true...),
    ...
}
```
```